### DISCUSSION TITLE
Cypher QA Summary

### DISCUSSION THREAD

**Discussion by KingSkyLi** [TuGraph-family/tugraph-db#102] [Q&A]
本贴是Cypher的QA汇总贴，详见下方回答。
This post is Cypher's QA summary post, see the answer below for details.

**Comment by KingSkyLi** [comment]
中文：
Q：是否支持不定长边的条件查询？
   示例：

   ```cypher
    MATCH p=(v)-[e:acted_in|:rate*1..3]-(v2) WHERE id(v) IN [3937] AND e.stars = 3 RETURN p LIMIT 100
   ```
A：目前还不支持不定长边的过滤查询。目前的代替方案只能是分开写。上面的示例，就需要从 1 跳到 3 跳都写一遍。


English:
Q: Does it support conditional query with indefinite length?
   Example:

   ```cypher
    MATCH p=(v)-[e:acted_in|:rate*1..3]-(v2) WHERE id(v) IN [3937] AND e.stars = 3 RETURN p LIMIT 100
   ```
A: Filtering queries with indefinite length edges are not yet supported. The current alternative is to write them separately. In the above example, it is necessary to write all jumps from 1 to 3.

**Comment by KingSkyLi** [comment]
中文：
Q：如何查询最短路径，shortestPath 函数如何使用？
A：使用示例如下（示例图谱：MovieDemo）
   ```cypher
    MATCH (n1 {name:'Corin Redgrave'}),(n2 {name:'Liam Neeson'})
        CALL algo.allShortestPaths(n1,n2) YIELD nodeIds,relationshipIds,cost
            RETURN nodeIds,relationshipIds,cost
   ```
   详尽使用方案请参考官网文档[https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658658](https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658658)。
  
 English: 
Q: How to query the shortest path, how to use the shortestPath function?
A: The usage example is as follows (sample map: MovieDemo)
   ```cypher
    MATCH (n1 {name:'Corin Redgrave'}),(n2 {name:'Liam Neeson'})
        CALL algo.allShortestPaths(n1,n2) YIELD nodeIds, relationshipIds, cost
            RETURN nodeIds, relationshipIds, cost
   ```
   For a detailed usage plan, please refer to the official website document [https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658658](https://www.tugraph.org/doc?version=V3.3.0&id=10000000000658658 ).

**Comment by KingSkyLi** [comment]
中文：
Q：查询语句 Where 后使用 and 进行拼接查询速度较慢，语句应如何优化改进？
   示例：

   ```cypher
    MATCH (n1),(n2) CALL algo.allShortestPaths(n1,n2)
        YIELD nodeIds,relationshipIds,cost
            WHERE id(n1) IN [0] AND id(n2) IN [3938]
                RETURN nodeIds,relationshipIds,cost
   ```
A：目前 cypher 查询引擎正在优化中。现阶段语句改写可以通过 with 向下传递进行优化。
   示例：

   ```cypher
    MATCH (n1) where id(n1) in [0] with n1
    MATCH (n2) where id(n2) in [3938] with n1, n2
        CALL algo.allShortestPaths(n1,n2) YIELD nodeIds,relationshipIds,cost
            RETURN nodeIds,relationshipIds,cost
   ```

English:
Q: Using and after the query statement Where to splice the query is slow, how should the statement be optimized and improved?
   Example:

   ```cypher
    MATCH (n1),(n2) CALL algo.allShortestPaths(n1,n2)
        YIELD nodeIds, relationshipIds, cost
            WHERE id(n1) IN [0] AND id(n2) IN [3938]
                RETURN nodeIds, relationshipIds, cost
   ```
A: Currently the cypher query engine is being optimized. Statement rewriting at this stage can be optimized by passing down with with .
   Example:

   ```cypher
    MATCH (n1) where id(n1) in [0] with n1
    MATCH (n2) where id(n2) in [3938] with n1, n2
        CALL algo.allShortestPaths(n1,n2) YIELD nodeIds, relationshipIds, cost
            RETURN nodeIds, relationshipIds, cost
   ```

**Comment by KingSkyLi** [comment]
中文：
Q：如何查询任意跳的边？
A：使用`*..`
   示例：
   ```cypher
    MATCH p=(a)-[*..]-(b) WHERE id(a) IN [3] AND id(b) IN [19] RETURN p
   ```
English:
Q: How to query the edge of any jump?
A: Use `*..`
   Example:
   ```cypher
    Match p=(a)-[*..]-(b) WHERE id(a) IN [3] AND id(b) IN [19] RETURN p
   ```

**Comment by zhangwh807** [comment]
中文：
Q：MERGE语法报错（"BadRequest CypherException: Function not implemented yet: ExtractNodePattern at :166"），查询语句（MERGE (n1:domain {name:'root',id: 'root', tag: 'root'})）
A：现在对于有多个属性的这种Merge语法还没完全支持，可以参考这几种用法：
1.  MERGE (n:Person {name:'Zhugeliang'}) ON CREATE SET n.gender=1,n.birthyear=181 RETURN n.name
2.  MERGE (n:Person {name:'Liubei'}) ON MATCH SET n.birthyear=2010 RETURN n.birthyear

English：
Q: The “MERGE (n1:domain {name:'root',id: 'root', tag: 'root'})” statement reports an error："BadRequest CypherException: Function not implemented yet: ExtractNodePattern at :166"
A: The Merge syntax with multiple attributes is not fully supported, so here are a few ways to use it:
1.  MERGE (n:Person {name:'Zhugeliang'}) ON CREATE SET n.gender=1,n.birthyear=181 RETURN n.name
2.  MERGE (n:Person {name:'Liubei'}) ON MATCH SET n.birthyear=2010 RETURN n.birthyear

**Comment by zhangwh807** [comment]
中文：
Q：如何获取每个Cypher请求的耗时?
A：TuGraph的cli可以看到时间：
https://github.com/TuGraph-family/tugraph-db/blob/master/docs/zh-CN/source/6.utility-tools/6.tugraph-cli.md
如果是白屏前端，可以看接口调用的端到端时间

English:
For TuGraph's CLI, you can view timing information as described in the documentation here: 
https://github.com/TuGraph-family/tugraph-db/blob/master/docs/zh-CN/source/6.utility-tools/6.tugraph-cli.md
If it is a web frontend, you can see the end-to-end time of an API call.

**Comment by zhangwh807** [comment]
中文：
Q：多条件场景下，查询报错，语法如下：MATCH (tom:Person{name:"Andres",title:"Developer"}) RETURN tom
需要修改成：MATCH (tom:Person{name:"Andres"}) RETURN tom才能正确执行，麻烦看下是否不支持多条件查询，或者是否有其他的替代查询方式
A：可以尝试这样查询
MATCH (tom:Person) WHERE tom.name = 'Andres' AND tom.title = 'Developer' RETURN tom

English:
Q: When facing a multi-condition scenario, I get an error with the following syntax:
MATCH (tom:Person{name:"Andres",title:"Developer"}) RETURN tom
It needs to be modified to:
MATCH (tom:Person{name:"Andres"}) RETURN tom
to execute correctly. Please check if multi-condition queries are unsupported, or if there is an alternative query method.
A: You can try querying like this:
MATCH (tom:Person) WHERE tom.name = 'Andres' AND tom.title = 'Developer' RETURN tom
